controllers/machine/machine_controller.go

```diff
@@ -0,0 +1,194 @@
+	// if there are no events at all exit now
+	if len(eventList.Items) == 0 {
+		reqLogger.Info("No events")
+		return nil
+	}
+
+	var newestEvent *corev1.Event
+
+	for i := range eventList.Items {
+		event := &eventList.Items[i]
+		if event.Reason == "DrainRequeued" {
+			if strings.Contains(event.Message, "error when evicting pods") {
+				if newestEvent == nil {
+					newestEvent = event
+					continue
+				}
+				if newestEvent.LastTimestamp.Time.Before(event.LastTimestamp.Time) {
+					newestEvent = event
+					continue
+				}
+			}
+		}
+	}
+
+	return newestEvent
+}
+
+func parsePodsAndNamespacesFromEvent(reqLogger logr.Logger, event *corev1.Event) map[string]string {
+	re := regexp.MustCompile(`pods\/"([\w-]+)" -n "([\w-]+)"`)
+	matches := re.FindAllStringSubmatch(event.Message, -1)
+
+	// store the pod/namespace combinations with the pod as the key because
+	// the pod name _should_ generally be unique, where there may be multiple pods
+	// in each namespace
+	podNamespaces := map[string]string{}
+
+	for _, podMatch := range matches {
+		if len(podMatch) != 3 {
+			// I don't think this should ever happen, but this prevents trying to access indexes
+			// in the match slice that may not exist
+			reqLogger.Error(fmt.Errorf("Could not get the appropriate amount of matches"), "match", podMatch)
+			continue
+		}
+		// From the regex match we'll always get this podMatch slice with the following format:
+		// ['pods/"myPod-aaabbb" -n "namespace"', 'myPod-aaabbb', 'namespace']
+		podName := podMatch[1]
+		podNamespace := podMatch[2]
+
+		namespaceRe := regexp.MustCompile(`openshift-.*`)
+		// We don't generally care about openshift namespaces for customer metrics
+		if !namespaceRe.MatchString(podNamespace) {
+			podNamespaces[podName] = podNamespace
+		}
+	}
+	return podNamespaces
+}
+
+func (r *MachineReconciler) evaluateDeletingMachine(ctx context.Context, machine *machinev1beta1.Machine) (ctrl.Result, error) {
+	reqLogger := logf.FromContext(ctx).WithName(logName)
+
+	// Check Deleting Timestamp. If it's been less than 15m we don't care, requeue for 5m.
+	deletedTime := machine.GetDeletionTimestamp().Time
+	now := time.Now()
+	if !deletedTime.Before(now.Add(-15 * time.Minute)) {
+		reqLogger.Info("Machine was not deleted long enough ago. Requeueing after 5m.")
+		return utils.RequeueAfter(5 * time.Minute)
+	}
+
+	reqLogger.Info("Evaluating Deleting Machine")
+
+	machineEventList := &corev1.EventList{}
+	// we only want the events related to the machine that we're reconciling on
+	err := r.Client.List(ctx, machineEventList, &client.ListOptions{Namespace: "openshift-machine-api", FieldSelector: fields.SelectorFromSet(fields.Set{"involvedObject.name": machine.Name})})
+	if err != nil {
+		reqLogger.Error(err, "Unable to query events for machine")
+		return utils.RequeueWithError(err)
+	}
+
+	event := getMostRecentDrainFailedEvent(reqLogger, machineEventList)
```

Comment:
I'm wondering if we should check weather the `event.LastTimestamp > machine.DeletionTimestamp`...
Is it even possible that we find a drain failed event for a previous deletion attempt?

Reply:
Great idea. It would be possible, there's nothing saying you can't cordon a node, drain it, and then uncordon it. That wouldn't replace the node or machine, and then it would work again until it's deleted again.